### PR TITLE
[rllib] Move a3c implementation from examples/ to python/ray/rllib/

### PR DESCRIPTION
**Pull request by ericl**
This moves examples/a3c to the rllib dir, and also refactors it slightly to conform to the `ray.rllib.common.Algorithm` interface.

Notably, each train() call will return intermediate training results after 100 rollouts (configurable) have been processed. I don't think this affects the semantics of the algorithm, but it would be good to have a more careful review of the changes there.

cc @richardliaw @pcmoritz

### CODE REVIEW

File: python/ray/rllib/a3c/runner.py [hunk 1]
```diff
@@ -30,10 +30,11 @@ def process_rollout(rollout, gamma, lambda_=1.0):
 
   features = rollout.features[0]
   return Batch(batch_si, batch_a, batch_adv, batch_r, rollout.terminal,
-               features)
+               features, np.sum(rewards))
```
Comment (ericl) [inline python/ray/rllib/a3c/runner.py]: Had to add this in order to report mean reward as part of TrainingResult.

File: python/ray/rllib/a3c/runner.py [hunk 2]
```diff
@@ -90,26 +95,37 @@ def start_runner(self, sess, summary_writer):
     self.start()
 
   def run(self):
-    with self.sess.as_default():
-      self._run()
+    try:
+      with self.sess.as_default():
+        self._run()
+    except BaseException as e:
+      self.queue.put(e)
```
Comment (ericl) [inline python/ray/rllib/a3c/runner.py]: Propagates the exception to the driver.

File: python/ray/rllib/a3c/runner.py [hunk 3]
```diff
@@ -127,10 +143,13 @@ def env_runner(env, policy, num_local_steps, summary_writer, render):
       if render:
         env.render()
 
-      # Collect the experience.
-      rollout.add(last_state, action, reward, value_, terminal, last_features)
       length += 1
       rewards += reward
+      if length >= timestep_limit:
+        terminal = True
```
Comment (ericl) [inline python/ray/rllib/a3c/runner.py]: Previously these weren't marked as terminal, which I think could cause reward values to be incorrectly totaled later on.

**Comment by ericl** [comment]
Ping

**Comment by robertnishihara** [comment]
Looks good to me. Looks like linting is failing on Travis with one minor error.

**Comment by pcmoritz** [comment]
Yeah looks good! Let's fix the linting and merge it.

**Comment by ericl** [comment]
Lint fixed.

**Pull request merged**